File: subly_ios/Subly/Subly/UI/PaymentMethods/PaymentMethodsViewModel.swift
import Foundation
import Combine
import Observation

@Observable
final class PaymentMethodsViewModel {
    var uiState = PaymentMethodsUiState()

    private let paymentMethodRepository: PaymentMethodRepository
    private var cancellables = Set<AnyCancellable>()

    init(paymentMethodRepository: PaymentMethodRepository) {
        self.paymentMethodRepository = paymentMethodRepository
        loadPaymentMethods()
    }

    private func loadPaymentMethods() {
        paymentMethodRepository.getAllPaymentMethods()
            .receive(on: DispatchQueue.main)
            .sink { [weak self] methods in
                guard let self else { return }
                Task { @MainActor in
                    var withUsage: [PaymentMethodWithUsage] = []
                    for method in methods {
                        let count = await self.paymentMethodRepository.getSubscriptionCount(for: method.id)
                        withUsage.append(PaymentMethodWithUsage(paymentMethod: method, subscriptionCount: count))
                    }
                    self.uiState.paymentMethods = withUsage
                    self.uiState.isLoading = false
                    self.uiState.error = nil
                }
            }
            .store(in: &cancellables)
    }

    func requestDelete(_ item: PaymentMethodWithUsage) {
        if item.subscriptionCount > 0 {
            uiState.error = "Cannot delete: used by \(item.subscriptionCount) subscription\(item.subscriptionCount == 1 ? "" : "s")"
            return
        }
        uiState.deletingMethod = item
        uiState.showDeleteConfirmation = true
    }

    func dismissDelete() {
        uiState.deletingMethod = nil
        uiState.showDeleteConfirmation = false
    }

    func confirmDelete() {
        guard let item = uiState.deletingMethod else { return }
        uiState.isDeleting = true
        uiState.showDeleteConfirmation = false

        Task { @MainActor in
            do {
                try await paymentMethodRepository.deletePaymentMethod(id: item.paymentMethod.id)
                uiState.isDeleting = false
                uiState.deletingMethod = nil
            } catch {
                uiState.isDeleting = false
                uiState.deletingMethod = nil
                uiState.error = "Failed to delete: \(error.localizedDescription)"
            }
        }
    }

    func clearError() {
        uiState.error = nil
    }
}

File: subly_android/app/src/main/java/net/cynreub/subly/ui/payment/PaymentMethodsViewModel.kt
package net.cynreub.subly.ui.payment

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import net.cynreub.subly.domain.repository.PaymentMethodRepository
import java.util.UUID
import javax.inject.Inject

@HiltViewModel
class PaymentMethodsViewModel @Inject constructor(
    private val paymentMethodRepository: PaymentMethodRepository
) : ViewModel() {

    private val _uiState = MutableStateFlow(PaymentMethodsUiState())
    val uiState: StateFlow<PaymentMethodsUiState> = _uiState.asStateFlow()

    init {
        loadPaymentMethods()
    }

    private fun loadPaymentMethods() {
        viewModelScope.launch {
            paymentMethodRepository.getAllPaymentMethods()
                .catch { e ->
                    _uiState.value = PaymentMethodsUiState(
                        isLoading = false,
                        error = "Failed to load payment methods: ${e.message}"
                    )
                }
                .collectLatest { paymentMethods ->
                    // Get subscription count for each payment method
                    val paymentMethodsWithUsage = paymentMethods.map { paymentMethod ->
                        val count = try {
                            paymentMethodRepository.getSubscriptionCountForPaymentMethod(paymentMethod.id)
                        } catch (e: Exception) {
                            0 // Default to 0 if count fails
                        }
                        PaymentMethodWithUsage(paymentMethod, count)
                    }

                    _uiState.value = PaymentMethodsUiState(
                        paymentMethods = paymentMethodsWithUsage,
                        isLoading = false,
                        error = null
                    )
                }
        }
    }

    fun showDeleteDialog(paymentMethodWithUsage: PaymentMethodWithUsage) {
        _uiState.value = _uiState.value.copy(
            showDeleteDialog = true,
            deletingPaymentMethod = paymentMethodWithUsage
        )
    }

    fun dismissDeleteDialog() {
        _uiState.value = _uiState.value.copy(
            showDeleteDialog = false,
            deletingPaymentMethod = null
        )
    }

    fun onDeleteConfirm() {
        val paymentMethodWithUsage = _uiState.value.deletingPaymentMethod ?: return

        if (paymentMethodWithUsage.subscriptionCount > 0) {
            _uiState.value = _uiState.value.copy(
                showDeleteDialog = false,
                deletingPaymentMethod = null,
                error = "Cannot delete: used by ${paymentMethodWithUsage.subscriptionCount} subscription(s)"
            )
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(
                isDeleting = true,
                showDeleteDialog = false
            )

            try {
                paymentMethodRepository.deletePaymentMethodById(paymentMethodWithUsage.paymentMethod.id)
                _uiState.value = _uiState.value.copy(
                    isDeleting = false,
                    deletingPaymentMethod = null,
                    error = null
                )
                // Data will automatically refresh via Flow
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isDeleting = false,
                    deletingPaymentMethod = null,
                    error = "Failed to delete payment method: ${e.message}"
                )
            }
        }
    }

    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }

    fun refresh() {
        _uiState.value = _uiState.value.copy(isLoading = true)
        loadPaymentMethods()
    }
}
